### ISSUE TITLE
Disable a Refactoring

### ISSUE THREAD

**Issue by christophwille** [dotnet/roslyn#4032]
Provide a UI to allow users to selectively disable refactorings that are installed via 3rd party VSIXs.

**Comment by hellfirehd** [comment]
Sorry for resurrecting such an old issue.  But since it's still open... 

> Provide a UI to allow users to selectively disable refactorings that are installed via 3rd party VSIXs.

Not just 3rd Party! I'd like to disable some built in ones as well.  Specifically the ["Use discard '_'"](https://docs.microsoft.com/en-us/visualstudio/ide/reference/unused-values-parameters?view=vs-2019) refactoring.  I have already disabled the analyzer (`IDE0059`) but when I `Alt+Enter` on the line, the *Use discard '-'* is the first selection and I frequently trigger it accidentally.  This slows me down and is generally frustrating as it happens a lot.

In any event, I really want to be able to turn any refactoring on and off and if it was a **per solution** setting that would simply be amazing.

----

Regarding `IDE0059` specifically, I would expect VS/Roslyn to be smart enough to use the discard on it's own (and not clutter my code with _ everywhere) since I haven't assigned the result explicitly.

In my opinion, the following lines should NOT trigger the *Use discard '_'* refactoring regardless of `IDE0059`, but that's probably an issue to log somewhere else:
```csharp
services.AddSingleton<IClock>(SystemClock.Instance);
services.AddSingleton<IDateTimeZoneProvider, DateTimeZoneCache>();
services.AddSingleton<IDateTimeZoneSource, TzdbDateTimeZoneSource>();
```

Incidentally, disabling `IDE0059` also disables detection of redundant assignments which is unfortunate, but I can live with it as that occurs far less often.

**Comment by jmarolf** [comment]
I do think this is good idea. Just haven't gotten around to implementing it yet. Out of curiosity, where would you expect to configure this from? The lightbulb itself? Tools->Options? Some nee UI?

> Regarding IDE0059 specifically, I would expect VS/Roslyn to be smart enough to use the discard on it's own

Discard is about you showing that you _intended_ to ignore the result. It can be hard to discern if a return value was ignored on purpose or if it is an error. Adding a discard tells future generations that it was intentional. For the case you describe, we could consider shipping suppression analyzers so that APIs that have return values that are almost always ignored could be configured in some way, though honestly refactoring's are not suggestions they are just things you could do. I don't think it makes sense to not show valid things to do. I _do_ think we need to make sure that if there is a codefix that fixes an error or a warning its not superceded by a helpful refactoring.

**Comment by CyrusNajmabadi** [comment]
> Not just 3rd Party! I'd like to disable some built in ones as well. Specifically the "Use discard '_'" refactoring. 

This can be disabled:

![image](https://user-images.githubusercontent.com/4564579/78960039-a2ed3780-7aa1-11ea-83aa-b80865ad4729.png)

**Comment by hellfirehd** [comment]
@CyrusNajmabadi No, that doesn't turn it off.  It changes the behavior but the refactoring is still enabled. (Thus the "Refactoring Only".)

**Comment by hellfirehd** [comment]
> where would you expect to configure this from? The lightbulb itself? Tools->Options? Some nee UI?

Tools -> Options would be fine or something like the `.ruleset` files we have for the analyzers.

>It can be hard to discern if a return value was ignored on purpose or if it is an error. Adding a discard tells future generations that it was intentional.

If future generations need that help then we're in trouble. I'm not joking.  If the return value isn't used later in the method then clearly it's okay to ignore it.  There are a couple places that the discard makes sense:
```csharp
public Boolean TryRemove(T item) => _dictionary.TryRemove(item, out _);
...
services.Configure<CookiePolicyOptions>(options =>
{
  options.CheckConsentNeeded = _ => true;
});
...
smtp.ServerCertificateValidationCallback = (_, __, ___, ____) => true;  // <- So Ugly!
```
Here it is absolutely not needed and is just noise:
```csharp
_ = await SaveChangesAsync().ConfigureAwait(false);
...
_ = services.AddDataProtection()
            .PersistKeysToDbContext<AuthDbContext>();
```

> refactoring's are not suggestions they are just things you could do. I don't think it makes sense to not show valid things to do. 

Sure it does! That's the whole point of this ticket. I don't use the discard for unused assignments so everytime that refactoring pops up and I hit it accidentally, it's causing me frustration.

> I do think we need to make sure that if there is a codefix that fixes an error or a warning its not superceded by a helpful refactoring.

That would certainly be an improvement.

**Comment by CyrusNajmabadi** [comment]
> If future generations need that help then we're in trouble. I'm not joking. If the return value isn't used later in the method then clearly it's okay to ignore it.

That's not the purpose. The fact that it isn't used later may actually be a bug.  That's the point of being explicit here. To indicate if not using the return value is an intentional decision as opposed to an accidental omission. 

Regardless, I'm not getting the problem. Change the feature value. Now you won't be prompted for this in the task list or the editor.

**Comment by hellfirehd** [comment]
I had already changed it to the settings you show in your screen capture which resulted in the frustrating behavior: The *Suggestion* no longer appears but if I `Alt-Enter` on the line for a different reason (say Introduce Local...) the first option available is *Use discard '_'* which I often trigger accidentally.  Blame it on fat fingers, poor typing skills or too much caffeine, the fact is it happens and I'm not wrong to ask for the option to turn it off.

**Comment by CyrusNajmabadi** [comment]
> the first option available is Use discard '_' 

That's simply a bug.  This should be low-priority and should be at the bottom of teh list when in 'refactoring mode'.  That's trivially fixable.  Tagging @mavasani .

**Comment by mavasani** [comment]
@hellfirehd You want to disable IDE0058 and IDE0059 in your .editorconfig file and you will not even see these as refactorings:

```
[*.cs]
dotnet_diagnostic.IDE0058.severity = none
dotnet_diagnostic.IDE0059.severity = none
```

**Comment by hellfirehd** [comment]
@mavasani Thank you! That worked!  

FWIW, I only needed to disable `IDE0058`.  I'm not sure if `IDE0059` has an associated codefix/refactoring but in the situations that the *Use discard '_'* was appearing `IDE0059` seems to put nothing in the popup.

**Comment by allisonchou** [comment]
Notes from today's design meeting (4/13):
For now, we plan to add an .editorconfig option to disable refactorings by name. We may consider adding a Tools->Options disable option in the future based on user feedback.

**Comment by okarpov** [comment]
I need just refactor strings into constants but i can not do it because of endless messagebox "Gaithering Suggestions - 'ExtractMethodCodeRefactoringProvider'"

**Comment by CyrusNajmabadi** [comment]
@okarpov can you collect a trace and provide it to us?  Instructions here: https://github.com/dotnet/roslyn/blob/main/docs/wiki/Recording-performance-traces-with-PerfView.md

It's unlikely this is actually extract-method.  It's more likely that that is just calling into a lower level helper (that all refactorings would call) that is taking a long time.  

The trace will let us know definitively. Thanks!

**Comment by okarpov** [comment]
> @okarpov can you collect a trace and provide it to us? Instructions here: https://github.com/dotnet/roslyn/blob/main/docs/wiki/Recording-performance-traces-with-PerfView.md
> 
> It's unlikely this is actually extract-method. It's more likely that that is just calling into a lower level helper (that all refactorings would call) that is taking a long time.
> 
> The trace will let us know definitively. Thanks!

does not work
![Screenshot 2023-01-15 210922](https://user-images.githubusercontent.com/2629775/212561875-b21ffa8b-524f-4e87-8ffb-9527b31fe721.png)

**Comment by okarpov** [comment]
log from perfview:

Started with command line: "C:\temp\PerfView.exe" collect C:\temp\ReproTrace.etl -CircularMB:4096 -BufferSizeMB:256 -Merge:true -Zip:true -Providers:641d7f6c-481c-42e8-ab7e-d18dc5e5cb9e,*Microsoft-VisualStudio-Common,*RoslynEventSource,*StreamJsonRpc -ThreadTime -NoV2Rundown -NoNGenRundown
PerfView Version: 3.0.6  BuildDate: Mon, 07 Nov 2022 20:31:05 GMT
PerfView Start Time 1/15/2023 9:14:31 PM
Started: Collect
Warning: PdbScope not found at C:\temp\PerfViewExtensions\PdbScope.exe
Disabling the Image Size Menu Item.
Warning: ILSize not found at C:\temp\PerfViewExtensions\ILSize.dll
Disabling the IL Size Menu Item.
[Kernel Log: C:\temp\ReproTrace.kernel.etl]
Kernel keywords enabled: ThreadTime
Collection Dialog open.
Aborting tracing for sessions 'NT Kernel Logger' and 'PerfViewSession'.
Ensuring .NET Allocation profiler not installed.
Detected 64 bit system, removing 32 bit keys.
Completed: Collect   (Elapsed Time: 0.288 sec)
Exception Occurred: System.Runtime.InteropServices.COMException (0x800700AA): The requested resource is in use. (Exception from HRESULT: 0x800700AA)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at Microsoft.Diagnostics.Tracing.Session.TraceEventSession.EnableKernelProvider(Keywords flags, Keywords stackCapture)
   at PerfView.CommandProcessor.Start(CommandLineArgs parsedArgs)
   at PerfView.CommandProcessor.Collect(CommandLineArgs parsedArgs)
   at PerfView.MainWindow.<>c__DisplayClass23_0.<ExecuteCommand>b__0()
   at PerfView.StatusBar.<>c__DisplayClass22_0.<StartWork>b__0()
An exceptional condition occurred, see log for details.
Collection Dialog open.
Started: Collecting data C:\temp\ReproTrace.etl
[Kernel Log: C:\temp\ReproTrace.kernel.etl]
Kernel keywords enabled: ThreadTime
Completed: Collecting data C:\temp\ReproTrace.etl   (Elapsed Time: 0.212 sec)
Exception Occurred: System.Runtime.InteropServices.COMException (0x800700AA): The requested resource is in use. (Exception from HRESULT: 0x800700AA)
   at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at Microsoft.Diagnostics.Tracing.Session.TraceEventSession.EnableKernelProvider(Keywords flags, Keywords stackCapture)
   at PerfView.CommandProcessor.Start(CommandLineArgs parsedArgs)
   at PerfView.CommandProcessor.Collect(CommandLineArgs parsedArgs)
   at PerfView.MainWindow.<>c__DisplayClass23_0.<ExecuteCommand>b__0()
   at PerfView.StatusBar.<>c__DisplayClass22_0.<StartWork>b__0()
An exceptional condition occurred, see log for details.

**Comment by CyrusNajmabadi** [comment]
We addressed "Gaithering Suggestions - 'ExtractMethodCodeRefactoringProvider'" with a rewrite of Extract Method.

Prioritization/disabling of analyzers is available.  If individual problems are found at this point, please open new issues so we can explicitly target fixes for them.  Thanks.
